docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,235 @@
+
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+# Create a Katib client
+katib_client = KatibClient(namespace='kubeflow')
+
+# Run the tuning Experiment
+exp_name = "llm-hp-optimization-test"
+katib_client.tune(
+	name = exp_name,
+	# BERT model URI and type of Transformer to train it.
+	model_provider_parameters = HuggingFaceModelParams(
+		model_uri = "hf://google-bert/bert-base-cased",
+		transformer_type = transformers.AutoModelForSequenceClassification,
+	),
+	# Use 3000 samples from Yelp dataset.
+	dataset_provider_parameters = HuggingFaceDatasetParams(
+		repo_id = "yelp_review_full",
+		split = "train[:3000]",
+	),
+	# Specify HuggingFace Trainer parameters.
+	trainer_parameters = HuggingFaceTrainerParams(
+		training_parameters = transformers.TrainingArguments(
+			output_dir = "test_trainer",
+			save_strategy = "no",
+			eval_strategy = "epoch",
+			disable_tqdm = True,
+			log_level = "info",
+			learning_rate = katib.search.double(min=1e-05, max=5e-05),
+			per_device_train_batch_size = katib.search.categorical([8, 16, 32]),
+			num_train_epochs = katib.search.int(min=1, max=10),
+			weight_decay = katib.search.double(min=0.0, max=1.0),
+		),
+		# Set LoRA config to reduce number of trainable model parameters.
+		lora_config = LoraConfig(
+			r = katib.search.int(min=8, max=32),
+			lora_alpha = 8,
+			lora_dropout = 0.1,
+			bias = "none",
+		),
+	),	
+	objective_metric_name = "eval_accuracy", 
+	objective_type = "maximize", 
+	algorithm_name = "random",
+	max_trial_count = 50,
+	parallel_trial_count = 2,
+	resources_per_trial = DistributedResources(
+		num_workers = 4,
+		num_procs_per_worker = 2,
+		resources_per_worker = {
+			"gpu": 2,
+			"cpu": 1,
+			"memory": "10G",
+		},
+	),
+)
+
+# Get the best hyperparameters
+best_hps = katib_client.get_optimal_hyperparameters(exp_name)
+```
+
+## Implementation
+
+By passing the specified parameters, this API will automate hyperparameter optimization for LLMs. The implementation will focus on the following aspects:
+
+**Model and Dataset Management**: We will leverage the [storage_initializer](https://github.com/kubeflow/training-operator/tree/master/sdk/python/kubeflow/storage_initializer) from the Training Python SDK for seamless integration of pretrained models and datasets from platforms like HuggingFace and S3. This component manages downloading and storing pretrained models and datasets via a PersistentVolumeClaim (PVC), which is shared across containers, ensuring efficient access to the pretrained model and dataset without redundant downloads.
+
+**Hyperparameter Configuration**: Users specify training parameters and the hyperparameters to be optimized within `trainer_parameters`. The API will first traverse `trainer_parameters.training_parameters` and `trainer_parameters.lora_config` to identify tunable hyperparameters and set up their values for the Experiment and Trials. These parameters are then passed as `args` to the container spec of workers.
```

Comment:
@andreyvelich @johnugeorge I have investigated how Hugging Face and Optuna handle the conversion of hyperparameters. It appears that they first [set up the hyperparameters](https://github.com/huggingface/transformers/blob/1c68f2cafb4ca54562f74b66d1085b68dd6682f5/src/transformers/trainer.py#L1485) , then use PyTorch's `torch.distributed.broadcast_object_list` function to broadcast the hyperparameter values from the master to the worker nodes. During this process, they employ ["pickle.dumps"](https://github.com/huggingface/transformers/blob/5e89b335ab088b52e5dbb50c5368d944e9052ed4/src/transformers/integrations/integration_utils.py#L208) and ["pickle.loads"](https://github.com/huggingface/transformers/blob/5e89b335ab088b52e5dbb50c5368d944e9052ed4/src/transformers/integrations/integration_utils.py#L237) to serialize and deserialize the `TrainingArguments` when broadcasting the parameters. This approach is similar to the methods used in the Training Operator, where we use (['json.dumps'](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/sdk/python/kubeflow/training/api/training_client.py#L272) and ['json.loads'](https://github.com/kubeflow/training-operator/blob/06f2ceb2123eb6e1e9bd1d1cdcea244da56f2012/sdk/python/kubeflow/trainer/hf_llm_training.py#L161)). So I didn't find a better alternative than using string substitution.

For hyperparameter configuration, following the process used in HuggingFace and Optuna, I first set up the parameters for the Experiment and Trials, and then continued to use the string substitution method to pass the parameters as `args` in the container spec of workers.

Please review this and correct me if my understanding is incorrect.

Reply:
I see, thanks for exploring this @helenxie-bit! @johnugeorge Anything you think we can re-use in Katib ? I guess, since the substitution of HyperParameters is done on the controller side, we still need to pass special template in Trial spec.

Reply:
Thanks for the exploration @helenxie-bit